### PR TITLE
ci(e2e): e2e with playwright

### PR DESCRIPTION
**Pull request by subzero10**
## Status
**Ready**

## Description
I decided to give playwright a go and see how it performs versus our current integration tests.

- [x] Run tests on Browserstack from Playwright. Running on minimum and latest versions. Supported browsers: **chrome** and **edge**.
- [x] Run tests on local browser executables (shipped with playwright): Firefox, Chrome, Edge, Safari, Chromium and mobile browsers

### Karma vs Playwright
**Karma/Jasmine**

_Pros:_
- runs on Browserstack, supports all browsers and browser versions available on Browserstack

_Cons:_
- feels outdated
- tests are becoming more and more flaky
- lots of usage of setTimeout (especially since the switch to promises from the callback-based approach)
- difficult to debug

**Playwright**

_Pros:_
- modern api
- fast (even faster when not connected to remote browsers on browserstack)
- support for flaky tests
- local browser testing (installed through npm)
- browserstack support (though limited to chrome and edge)

_Cons:_
- recommended approach is to test only on latest browser versions, but we also want to test on older versions

## Todos
- [x] Migrate all tests to playwright
- [x] Browserstack integration
- [x] Karma/Jasmine vs Playwright comparison
- [x] README.md - explain the minimal architecture (`server.js`)

### CODE REVIEW

**Comment by subzero10** [comment]
Tests are passing, but we now have "2 flaky tests" tagged from playwright.
I will look at them separately.

File: packages/js/test/e2e/global-setup.ts
Comment (BethanyBerkowitz) [inline packages/js/test/e2e/global-setup.ts:28]: Calling `sleep` like this is cannot be the right way to do this. But I see that you took it from the example project that browserstack/playwright put out... so not your fault! But yuck 🤦‍♀️ 

Maybe include a comment showing where the example code came from!

https://github.com/browserstack/node-js-playwright-browserstack/blob/main/global-setup.js

File: packages/js/test/e2e/global-setup.ts
Comment (BethanyBerkowitz) [inline packages/js/test/e2e/global-setup.ts:20]: Can you log out this error? I got this message but it didn't come with any actionable information.

**Review by BethanyBerkowitz — COMMENTED** [review]
This is exciting! I'm not done testing it out, but I wanted to post what I had so far. 

I did run into a problem in testing, which is that if you `ctrl+C` out of running the tests, it leaves `bsLocal` running, for me it was on port 45690. Then if you try running them again, Browserstack can't start because there's already something running on the port.

**Comment by subzero10** [comment]
> I did run into a problem in testing, which is that if you ctrl+C out of running the tests, it leaves bsLocal running, for me it was on port 45690. Then if you try running them again, Browserstack can't start because there's already something running on the port.

Fixed!

File: packages/js/test/e2e/browsers.ts
Comment (subzero10) [reply packages/js/test/e2e/browsers.ts:5]: No, nice catch!

File: packages/js/test/e2e/global-setup.ts
Comment (subzero10) [reply packages/js/test/e2e/global-setup.ts:28]: I decided to improve it :)
No more `sleep`!

File: packages/js/test/e2e/browserstack.config.ts
Comment (subzero10) [reply packages/js/test/e2e/browserstack.config.ts:18]: It's a copy/paste from [here](https://github.com/browserstack/typescript-playwright-browserstack/blob/main/browserstack.config.ts#L41C8-L41C8). Even though, there's a bug on this [line](https://github.com/browserstack/typescript-playwright-browserstack/blob/fd4cc68fcd3615a9041228ac3bfae37371afb070/browserstack.config.ts#L41C8-L41C8) 🤦

File: packages/js/test/e2e/global-setup.ts
Comment (subzero10) [reply packages/js/test/e2e/global-setup.ts:20]: With "log out", you mean to log the `err` itself, right? Done!

File: packages/js/test/e2e/browserstack.config.ts
Comment (subzero10) [reply packages/js/test/e2e/browserstack.config.ts]: I refactored the code, it's much cleaner now, thanks!

File: packages/js/test/e2e/browsers.ts
Comment (subzero10) [reply packages/js/test/e2e/browsers.ts]: Done!

**Comment by subzero10** [comment]
@BethanyBerkowitz Thanks for the detailed review. I went through all your comments. Feel free to go for a new review round :)

**Review by joshuap — APPROVED** [review]
Love it! Especially the pros/cons writeup—thanks for that. I'm sold on playwright. :)

tbh, I'd be OK with testing latest browsers if we ever had to compromise, but it's cool that we can also do the minimum versions. 👌

**Comment by subzero10** [comment]
> tbh, I'd be OK with testing latest browsers if we ever had to compromise, but it's cool that we can also do the minimum versions. 👌

💯

**Comment by subzero10** [comment]
I'll wait for @BethanyBerkowitz's review before merging, since there were more than a few changes since the first review 😄

File: packages/js/test/e2e/global-setup.ts
Comment (BethanyBerkowitz) [reply packages/js/test/e2e/global-setup.ts:28]: Thank you! I knew this code must have been copy/pasted from somewhere because your code is much better than this 😆

**Review by BethanyBerkowitz — APPROVED** [review]
Yay, thanks so much @subzero10!

I still get test failures when running locally, but since you've got a separate issue open for flakiness I think this one is ready to go.

**Comment by subzero10** [comment]
> I still get test failures when running locally, but since you've got a separate issue open for flakiness I think this one is ready to go.

I'm curious, have you tried running the tests only with playwright (i.e. disable the browserstack browsers and unset any browserstack env variables)?

**Comment by BethanyBerkowitz** [comment]
> > I still get test failures when running locally, but since you've got a separate issue open for flakiness I think this one is ready to go.
> 
> I'm curious, have you tried running the tests only with playwright (i.e. disable the browserstack browsers and unset any browserstack env variables)?

Mmm yeah I still get a bunch of failures -- 36 on my most recent run 🤔 
```
36 failed
    [chromium] › integration.spec.ts:392:7 › Browser Integration › it sends user feedback for notice on submit 
    [chromium] › integration.spec.ts:438:7 › Browser Integration › it closes user feedback form on cancel 
    [firefox] › integration.spec.ts:330:7 › Browser Integration › it shows user feedback form 
    [firefox] › integration.spec.ts:359:7 › Browser Integration › it shows user feedback form with custom labels 
    [firefox] › integration.spec.ts:392:7 › Browser Integration › it sends user feedback for notice on submit 
    [firefox] › integration.spec.ts:438:7 › Browser Integration › it closes user feedback form on cancel 
    [webkit] › integration.spec.ts:392:7 › Browser Integration › it sends user feedback for notice on submit 
    [webkit] › integration.spec.ts:438:7 › Browser Integration › it closes user feedback form on cancel 
    [Mobile Chrome] › integration.spec.ts:392:7 › Browser Integration › it sends user feedback for notice on submit 
    [Mobile Chrome] › integration.spec.ts:438:7 › Browser Integration › it closes user feedback form on cancel 
    [Mobile Safari] › integration.spec.ts:392:7 › Browser Integration › it sends user feedback for notice on submit 
    [Microsoft Edge] › integration.spec.ts:52:7 › Browser Integration › it logs browser type and version 
    [Microsoft Edge] › integration.spec.ts:58:7 › Browser Integration › it notifies Honeybadger of unhandled exceptions 
    [Microsoft Edge] › integration.spec.ts:66:7 › Browser Integration › it notifies Honeybadger manually 
    [Microsoft Edge] › integration.spec.ts:76:7 › Browser Integration › it reports multiple errors in the same process 
    [Microsoft Edge] › integration.spec.ts:91:7 › Browser Integration › it sends console breadcrumbs 
    [Microsoft Edge] › integration.spec.ts:105:7 › Browser Integration › it sends string value console breadcrumbs when null 
    [Microsoft Edge] › integration.spec.ts:119:7 › Browser Integration › it sends click breadcrumbs 
    [Microsoft Edge] › integration.spec.ts:136:7 › Browser Integration › it sends XHR breadcrumbs for relative paths 
    [Microsoft Edge] › integration.spec.ts:161:7 › Browser Integration › it sends XHR breadcrumbs for absolute paths 
    [Microsoft Edge] › integration.spec.ts:186:7 › Browser Integration › it sends fetch breadcrumbs 
    [Microsoft Edge] › integration.spec.ts:204:7 › Browser Integration › it sends navigation breadcrumbs 
    [Microsoft Edge] › integration.spec.ts:222:7 › Browser Integration › it sends notify breadcrumbs 
    [Microsoft Edge] › integration.spec.ts:241:7 › Browser Integration › it sends window.onerror breadcrumbs 
    [Microsoft Edge] › integration.spec.ts:260:7 › Browser Integration › it skips onunhandledrejection when already sent 
    [Microsoft Edge] › integration.spec.ts:277:7 › Browser Integration › it sends window.onunhandledrejection breadcrumbs when rejection is an Error 
    [Microsoft Edge] › integration.spec.ts:308:7 › Browser Integration › it skips window.onunhandledrejection breadcrumbs when rejection is not Error 
    [Microsoft Edge] › integration.spec.ts:330:7 › Browser Integration › it shows user feedback form 
    [Microsoft Edge] › integration.spec.ts:359:7 › Browser Integration › it shows user feedback form with custom labels 
    [Microsoft Edge] › integration.spec.ts:392:7 › Browser Integration › it sends user feedback for notice on submit 
    [Microsoft Edge] › integration.spec.ts:438:7 › Browser Integration › it closes user feedback form on cancel 
    [Microsoft Edge] › integration.spec.ts:470:7 › Web Worker Integration › it creates worker 
    [Google Chrome] › integration.spec.ts:330:7 › Browser Integration › it shows user feedback form 
    [Google Chrome] › integration.spec.ts:359:7 › Browser Integration › it shows user feedback form with custom labels 
    [Google Chrome] › integration.spec.ts:392:7 › Browser Integration › it sends user feedback for notice on submit 
    [Google Chrome] › integration.spec.ts:438:7 › Browser Integration › it closes user feedback form on cancel 
  111 passed (3.9m)
```

**Comment by subzero10** [comment]
That's very strange. Which nodejs version are you using?

**Comment by BethanyBerkowitz** [comment]
> That's very strange. Which nodejs version are you using?

Looks like I was on 16.13.0, but I also get a bunch of failures on 20.3.1.

**Comment by subzero10** [comment]
> > That's very strange. Which nodejs version are you using?
> 
> Looks like I was on 16.13.0, but I also get a bunch of failures on 20.3.1.

FYI, I'm running on node 16 as well.

**Comment by subzero10** [comment]
> @subzero10 I tried running the playwright tests just in Firefox using `--ui` so I could get a better sense of what's going on. It looks like for some reason the user feedback form is added to the DOM twice.
> 
> Happy to look further into this, but I figured you have much more knowledge than me in this area and may have a sense of what's wrong?
> 
> Screen Shot 2023-09-18 at 11 08 13 AM

Thanks for the insight! Yes, I can look into it! I remember adding a safeguard about this but it may not be working all the time.

**Comment by subzero10** [comment]
OK, I know why this _could_ happen, but I don't understand why.
For the tests, I had to disable the safeguard (`isUserFeedbackScriptUrlAlreadyVisible`) because I inject the feedback form script in a different way:
```
const handle = await page.evaluateHandle(_ => {
      // DISABLE SAFEGUARD AND SCRIPT LOADING FUNCTIONS
      window.Honeybadger.isUserFeedbackScriptUrlAlreadyVisible = () => false
      window.Honeybadger.appendUserFeedbackScriptTag = () => { }
      // ... ...
})
// ... ...

// LOAD SCRIPT USING PLAYWRIGHT
const relativePath = '../../dist/browser/honeybadger-feedback-form.js'
await page.addScriptTag({
      path: resolve(__dirname, relativePath)
})
// ... ...
```
However, I don't see why the `addScriptTag` would be called twice, or result in the script being added twice.
The only theory that could come to mind is that the `page` for the current test is the same as the `page` from the previous test, that already had the feedback form script loaded. But that can't be true, because each test is [isolated with a fresh page](https://playwright.dev/docs/writing-tests#test-isolation). By the way, I can't reproduce this behavior 🤷‍♂️.

**Comment by BethanyBerkowitz** [comment]
> OK, I know why this _could_ happen, but I don't understand why. For the tests, I had to disable the safeguard (`isUserFeedbackScriptUrlAlreadyVisible`) because I inject the feedback form script in a different way:
> 
> ```
> const handle = await page.evaluateHandle(_ => {
>       // DISABLE SAFEGUARD AND SCRIPT LOADING FUNCTIONS
>       window.Honeybadger.isUserFeedbackScriptUrlAlreadyVisible = () => false
>       window.Honeybadger.appendUserFeedbackScriptTag = () => { }
>       // ... ...
> })
> // ... ...
> 
> // LOAD SCRIPT USING PLAYWRIGHT
> const relativePath = '../../dist/browser/honeybadger-feedback-form.js'
> await page.addScriptTag({
>       path: resolve(__dirname, relativePath)
> })
> // ... ...
> ```
> 
> However, I don't see why the `addScriptTag` would be called twice, or result in the script being added twice. The only theory that could come to mind is that the `page` for the current test is the same as the `page` from the previous test, that already had the feedback form script loaded. But that can't be true, because each test is [isolated with a fresh page](https://playwright.dev/docs/writing-tests#test-isolation). By the way, I can't reproduce this behavior 🤷‍♂️.

Well, I'm not sure what you did in https://github.com/honeybadger-io/honeybadger-js/pull/1190 that could fix my problem, but the Firefox tests no longer error for me!

**Pull request merged**
